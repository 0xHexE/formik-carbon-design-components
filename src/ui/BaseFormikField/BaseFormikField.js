import PropTypes from 'prop-types';
import React from 'react';
import { useField } from 'formik';
import {
  Checkbox,
  DatePicker,
  Dropdown,
  RadioButtonGroup,
  Search,
} from 'carbon-components-react';

function BaseFormikField({ renderField: RenderField, ...props }) {
  const [field, meta] = useField(props);
  let extraRender = null;

  const finalOverrider = {};

  if (RenderField === Checkbox || RenderField === RadioButtonGroup) {
    finalOverrider.onChange = (e) => {
      field.onChange({
        target: { value: e, id: props.id, name: props.name },
      });
    };
  } else if (RenderField === Search) {
    props.defaultValue = props.defaultValue || props.value || meta.value || '';
    props.value = props.defaultValue;
    props.placeHolderText = props.placeHolderText || props.placeholder;
  } else if (RenderField === Checkbox) {
    finalOverrider.invalidText = null;
    extraRender = <p>{meta.error}</p>;
  } else if (RenderField === DatePicker) {
    finalOverrider.onChange = ([first, second]) => {
      if (props.datePickerType === 'single') {
        field.onChange({
          target: { value: first, id: props.id, name: props.name },
        });
      } else {
        field.onChange({
          target: { value: [first, second], id: props.id, name: props.name },
        });
      }
    };
  }

  const invalid = !!meta.error && meta.touched;

  const requiredInvalidString =
    RenderField === Checkbox ||
    RenderField === Search ||
    RenderField === DatePicker;

  // Here invalid text is not applicable for some components
  if (!requiredInvalidString) {
    finalOverrider.invalidText = meta.error;
  }

  return (
    <>
      <RenderField
        invalid={requiredInvalidString ? invalid.toString() : invalid}
        {...props}
        {...field}
        {...finalOverrider}
      />
      {extraRender}
    </>
  );
}

BaseFormikField.propTypes = {
  datePickerType: PropTypes.any,
  defaultValue: PropTypes.any,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  placeHolderText: PropTypes.any,
  placeholder: PropTypes.any,
  renderField: PropTypes.any.isRequired,
  value: PropTypes.any,
};

export default BaseFormikField;
