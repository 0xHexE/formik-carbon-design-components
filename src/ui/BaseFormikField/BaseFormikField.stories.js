import React from 'react';
import { getConfig } from '../../../scripts/storybook/storyConfig';
import BaseFormikField from './BaseFormikField';
import FormikForm from '../../utils/formik-form';
import { Form } from 'formik';
import {
  Checkbox,
  TextArea,
  TextInput,
  Search,
  DatePicker,
  RadioButtonGroup,
  RadioButton,
  DatePickerInput,
  Dropdown,
} from 'carbon-components-react';

export default getConfig({
  title: 'BaseFormikField',
  component: BaseFormikField,
});

export function Basic() {
  return (
    <FormikForm>
      {(props) => {
        return (
          <Form>
            <BaseFormikField
              name="email"
              id="email"
              labelText="Email"
              placeholder="Email"
              renderField={TextInput}
            />
            <BaseFormikField
              name="Information"
              id="information"
              labelText="Information"
              placeholder="Email"
              renderField={TextArea}
            />
            <BaseFormikField
              renderField={Checkbox}
              labelText="Checked"
              id="ch1"
              name="checkbox"
            />
            <BaseFormikField
              renderField={RadioButtonGroup}
              labelText="Checked"
              value={'search'}
              id="ch1"
              name="r1">
              <RadioButton labelText="Radio" value="radio-1" id="radio-1" />
            </BaseFormikField>
            <BaseFormikField
              renderField={Search}
              labelText="Checked"
              value={''}
              placeholder={'Search'}
              id="ch112"
              name="search"
            />
            <BaseFormikField
              id="data-picker"
              datePickerType="single"
              dateFormat="m/d/Y"
              renderField={DatePicker}
              name="date-picker">
              <DatePickerInput id="test" labelText="Date" />
            </BaseFormikField>
            <BaseFormikField
              id="data-picker"
              datePickerType="range"
              dateFormat="m/d/Y"
              renderField={DatePicker}
              name="date-picker2">
              <DatePickerInput id="test-dpi1" labelText="Date" />
              <DatePickerInput id="test-dpi2" labelText="Date" />
            </BaseFormikField>
            <BaseFormikField
              id={'drop-down'}
              name={'drop-down'}
              renderField={Dropdown}
              items={[{ text: 'Test1', value: 'test1' }]}
              itemToString={item => (item ? item.text : '')}
            />
            {/* eslint-disable-next-line react/prop-types */}
            {JSON.stringify(props.values, null, 2)}
          </Form>
        );
      }}
    </FormikForm>
  );
}
