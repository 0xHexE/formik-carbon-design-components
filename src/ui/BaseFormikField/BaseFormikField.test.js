/* eslint-disable no-console */
import { render } from '@testing-library/react';
import React from 'react';
import ReactDOM from 'react-dom';
import BaseFormikField from './BaseFormikField';
import FormikForm from '../../utils/formik-form';
import { Form } from 'formik';
import { mount } from 'enzyme';
import { TextInput } from 'carbon-components-react';

function generateElement() {
  return (
    <FormikForm>
      {() => {
        return (
          <Form>
            <BaseFormikField
              name="email"
              id="email"
              labelText="Email"
              placeholder="Email"
              renderField={TextInput}
            />
          </Form>
        );
      }}
    </FormikForm>
  );
}

describe('TextArea', () => {
  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(generateElement(), div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('render email in it', () => {
    const { getByText } = render(generateElement());
    getByText('Email');
  });

  test('match with snapshot', () => {
    const wrapper = mount(generateElement());
    expect(wrapper.html()).toMatchSnapshot();
  });
});
