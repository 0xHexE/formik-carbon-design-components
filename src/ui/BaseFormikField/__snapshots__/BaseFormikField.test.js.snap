// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea match with snapshot 1`] = `"<form action=\\"#\\"><div class=\\"bx--form-item bx--text-input-wrapper\\"><label for=\\"email\\" class=\\"bx--label\\">Email</label><div class=\\"bx--text-input__field-wrapper\\"><input id=\\"email\\" placeholder=\\"Email\\" type=\\"text\\" class=\\"bx--text-input bx--text__input\\" title=\\"Email\\" name=\\"email\\" value=\\"\\"></div></div></form>"`;
